toolkit/smart-contracts/plutus-data/src/key_value.rs

```diff
@@ -0,0 +1,82 @@
+use crate::{DataDecodingError, DecodingResult};
+use cardano_serialization_lib::{PlutusData, PlutusList};
+use sidechain_domain::byte_string::ByteString;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct KeyValueDatum {
+	pub key: String,
+	pub value: ByteString,
+}
+
+impl TryFrom<PlutusData> for KeyValueDatum {
+	type Error = DataDecodingError;
+	fn try_from(datum: PlutusData) -> DecodingResult<Self> {
+		let data_iter = datum.as_list().ok_or_else(|| DataDecodingError {
+			datum: datum.clone(),
+			to: "KeyValueDatum".into(),
+			msg: "Expected [String, Vec<u8>]".into(),
+		})?;
+		let key = data_iter.get(0).as_bytes().ok_or_else(|| DataDecodingError {
+			datum: datum.clone(),
+			to: "KeyValueDatum".into(),
+			msg: "Expected [String, Vec<u8>]".into(),
+		})?;
+		let value = data_iter.get(1).as_bytes().ok_or_else(|| DataDecodingError {
+			datum: datum.clone(),
+			to: "KeyValueDatum".into(),
+			msg: "Expected [String, Vec<u8>]".into(),
+		})?;
+		Ok(Self { key: String::from_utf8(key).unwrap(), value: value.into() })
```

Comment:
This unwrap deserves to be replaced with `map_error(...)?`. At this point it reading data from the external source.